### PR TITLE
Use gpgme for gpg detector, misc style fixes

### PR DESCRIPTION
**Pull request by mochaaP**
use gpgme for gpg detector
move ssh socket to ~ suffix

### CODE REVIEW

**Comment by mochaaP** [comment]
the hmac watcher could be migrated to use ykpers, but i don't have a setup to test for that

File: detector/gpg.go [hunk 1]
```diff
@@ -27,64 +28,52 @@ func WatchGPG(gpgPubringPath string, requestGPGCheck chan bool) {
...
-					return true
-				})
-
-				for isGPGCardBusy() || isGPGCardBusy() {
-					// wait...
-				}
+			return nil
+		})
+		t.Stop()
+		response <- err
```
Comment (maximbaz) [inline detector/gpg.go:56]:
```suggestion
		if !t.Stop() {
			response <- err
		}
```

If you stop the timer, there will be nobody to consume the message from `response`, so this send will block forever.

File: detector/hmac.go [hunk 2]
```diff
@@ -2,25 +2,26 @@ package detector
...
 )
 
 // WatchHMAC watches when YubiKey is waiting for a touch on a HMAC request
 func WatchHMAC(notifiers *sync.Map) {
 	devicesEvents := initInotifyWatcher("HMAC", "/dev", notify.Create, notify.Remove)
 	defer notify.Stop(devicesEvents)
 
-	yubikeyHidrawDevices := strset.New()
-	if devices, err := ioutil.ReadDir("/dev"); err == nil {
+	yubikeyHidrawDevices := mapset.NewSet()
```
Comment (maximbaz) [inline detector/hmac.go:23]: It was painful to review a pull request that combines lots of unrelated style changes, library replacements, and actual new feature all in one big patch. It would be very appreciated if you split this next time, if not in separate PRs, at least in separate commits.

**Review by maximbaz — CHANGES_REQUESTED** [review]
Thanks for the PR! It's a very interesting idea to use gpgme, it certainly makes it somewhat less hacky :grin:

I only have a single comment, a proposal to fix the app occasionally hang.

**Comment by maximbaz** [comment]
> the hmac watcher could be migrated to use ykpers, but i don't have a setup to test for that

Admittedly I don't use it myself either, but I believe it's as easy to set it up as to follow this:

```sh
$ ykman config usb -e OTP # enable OTP if you have it disabled
$ ykman otp chalresp --touch --generate 2 # generate a new random key in slot 2
$ ykman otp calculate 2 # when asked, give it any value (e.g. 00), then this will require touch
```

Interestingly, the current code does not detect the need to touch yubikey (for me). I don't know if it regressed (nobody opened an issue!), a quirk of my personal computer, or an issue in the setup above. But if you want to try it out, give it a go :grin:

File: detector/ssh.go [hunk 3]
```diff
@@ -27,7 +27,7 @@ func WatchSSH(requestGPGCheck chan bool, exits *sync.Map) {
 		return
 	}
 
-	originalSocketFile := socketFile + ".original"
+	originalSocketFile := socketFile + "~"
```
[the remark this answers is not in the document]
Comment (mochaaP) [reply detector/ssh.go:30]: the tilde is the common suffix for backup or temporary copies of files.
<https://unix.stackexchange.com/questions/6781/what-does-mean>

File: detector/hmac.go (hunk 2 above)
Comment (mochaaP) [reply detector/hmac.go:23]: Sorry for that! I was trying to update golang version, get this into the official Fedora repository at the same time, and this is the only working replacement for that dependency. 🤕

File: detector/ssh.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (maximbaz) [reply detector/ssh.go:30]: Yep - in our case it's neither of those. It's not a copy, it's an actively used operational file, and it sucks if you accidentally remove it, as everything will break :slightly_smiling_face: The file that we create is a proxy, that still sends every single request to the `.original` file.

[the remark this answers is not in the document]
Comment (mochaaP) [reply detector/ssh.go:30]: sensible. would it be okay to use an inotify based approach?

[the remark this answers is not in the document]
Comment (maximbaz) [reply detector/ssh.go:30]: Maybe I should ask first why does the file name matter for you? :slightly_smiling_face: Does it cause some troubles with regards to packaging?

I don't mind trying inotify-based approach, but I also don't mind changing the name if there is any other reason, I thought you simply wanted tilde for aesthetics reasons :upside_down_face:

[the remark this answers is not in the document]
Comment (mochaaP) [reply detector/ssh.go:30]: i think i'm fine with that now! will commit a change later

File: detector/hmac.go (hunk 2 above)
Comment (maximbaz) [reply detector/hmac.go:23]: Writing this from my Fedora: thanks! :pray: :grin: 

If you need any other change in the repo to make your life easier, do not hesitate to reach out or send PRs :wink:

**Comment by mochaaP** [comment]
revised with the suggestions.

**Comment by mochaaP** [comment]
ssh watcher could potentially use the sock_diag subsystem of netlink, but i couldn't find a good enough wrapper for that in golang.

**Review by maximbaz — APPROVED** [review]

**Comment by maximbaz** [comment]
FYI I doubled the timings, because I experienced sporadic false positives when executing `gpg sign` operations, which do not require a touch, but just because they are slow, they triggered yubikey-touch-detector's notification.

I also cut 1.11.0 release, so you can just take the latest tag for packaging :wink:

File: go.mod [hunk 4]
```diff
@@ -3,16 +3,18 @@ module github.com/maximbaz/yubikey-touch-detector
...
-	github.com/rjeczalik/notify v0.9.2
-	github.com/scylladb/go-set v1.0.2
-	github.com/sirupsen/logrus v1.8.1
+	github.com/proglottis/gpgme v0.1.3
+	github.com/rjeczalik/notify v0.9.3
+	github.com/sirupsen/logrus v1.9.3
 	github.com/vtolstov/go-ioctl v0.0.0-20151206205506-6be9cced4810
 )
 
+replace github.com/proglottis/gpgme => github.com/mcha-forks/gpgme v0.1.4-0.20230930202035-21b7c62d7377
```
Comment (brunnre8) [inline go.mod:15]: Why exactly are you pulling in your fork here rather than the original?

Comment (mochaaP) [reply go.mod:15]: Because https://github.com/proglottis/gpgme/pull/41 didn't get merged. :)
